include/sparrow/allocator.hpp

```diff
@@ -0,0 +1,247 @@
+        and std::equality_comparable<std::remove_cvref_t<T>>
+        and requires(std::remove_cvref_t<T>& alloc,
+                     typename std::remove_cvref_t<T>::value_type* p,
+                     std::size_t n)
+        {
+            typename std::remove_cvref_t<T>::value_type;
+            { alloc.allocate(n) } -> std::same_as<typename std::remove_cvref_t<T>::value_type*>;
+            { alloc.deallocate(p, n) } -> std::same_as<void>;
+        };
+
+    /*
+     * Type erasure class for allocators. This allows to use any kind of allocator
+     * (standard, polymorphic) without having to expose it as a template parameter.
+     *
+     * @tparam T value_type of the allocator
+     * @tparam DA default allocator, instantiated when calling the default constructor
+     */
+    template <class T, class DA = std::allocator<T>>
+    class any_allocator
+    {
+        template <class A>
+        static constexpr bool is_value_stored = 
+            std::is_same_v<A, std::allocator<T>> || std::is_same_v<A, std::pmr::polymorphic_allocator<T>>;
+
+    public:
+
+        using value_type = T;
+
+        any_allocator();
+        any_allocator(const any_allocator& rhs);
+        any_allocator(any_allocator&&) = default;
+
+        any_allocator& operator=(const any_allocator& rhs) = delete;
+        any_allocator& operator=(any_allocator&& rhs) = delete;
+
+        template <class A>
+        any_allocator(A&& alloc)
+        requires (not std::same_as<std::remove_cvref_t<A>, any_allocator>
+                and sparrow::allocator<A>)
+            : m_storage(make_storage(std::forward<A>(alloc)))
+        {
+        }
+
+        [[nodiscard]] T* allocate(std::size_t n);
+        void deallocate(T* p, std::size_t n);
+        
+        any_allocator select_on_container_copy_construction() const;
+
+        bool equal(const any_allocator& rhs) const; 
+
+    private:
+
+        struct interface
+        {
+            [[nodiscard]] virtual T* allocate(std::size_t) = 0;
+            virtual void deallocate(T*, std::size_t) = 0;
+            virtual std::unique_ptr<interface> clone() const = 0;
+            virtual bool equal(const interface&) const = 0;
+            virtual ~interface() = default;
+        };
+
+        template <allocator A>
+        struct impl : interface
+        {
+            A m_alloc;
+
+            explicit impl(A alloc) : m_alloc(std::move(alloc)) {}
+
+            [[nodiscard]] T* allocate(std::size_t n) override
+            {
+                return m_alloc.allocate(n);
+            }
+
+            void deallocate(T* p, std::size_t n) override
+            {
+                m_alloc.deallocate(p, n);
+            }
+
+            std::unique_ptr<interface> clone() const override
+            {
+                return std::make_unique<impl<A>>(m_alloc);
+            }
+
+            bool equal(const interface& rhs) const override
+            {
+                if (auto* p = dynamic_cast<const impl<A>*>(&rhs))
```

Comment:
Invoking `dynamic_cast` is super expensive and must be avoided at all costs.
Could a solution using `std::type_index(typeid(T))` work?